serving/src/main/java/feast/serving/service/OnlineServingService.java

```diff
@@ -0,0 +1,162 @@
+          retriever.getOnlineFeatures(entityRows, featureSetRequests);
+
+      for (var fsIdx = 0; fsIdx < featureRows.size(); fsIdx++) {
+        List<FeatureRow> featureRowsForFs = featureRows.get(fsIdx);
+        FeatureSetRequest featureSetRequest = featureSetRequests.get(fsIdx);
+        Map<String, FeatureReference> featureNames =
+            featureSetRequest.getFeatureReferences().stream()
+                .collect(
+                    Collectors.toMap(
+                        FeatureReference::getName, featureReference -> featureReference));
+        for (var entityRowIdx = 0; entityRowIdx < entityRows.size(); entityRowIdx++) {
+          FeatureRow featureRow = featureRowsForFs.get(entityRowIdx);
+          EntityRow entityRow = entityRows.get(entityRowIdx);
+          if (isStale(featureSetRequest, entityRow, featureRow)) {
+            featureSetRequest
+                .getFeatureReferences()
+                .parallelStream()
+                .forEach(
+                    ref -> {
+                      staleKeyCount
+                          .labels(
+                              featureSetRequest.getSpec().getProject(),
+                              String.format("%s:%d", ref.getName(), ref.getVersion()))
+                          .inc();
+                      featureValuesMap
+                          .get(entityRow)
+                          .put(RefUtil.generateFeatureStringRef(ref), Value.newBuilder().build());
+                    });
+
+          } else {
+            featureSetRequest
+                .getFeatureReferences()
+                .parallelStream()
+                .forEach(
+                    ref ->
+                        requestCount
+                            .labels(
+                                featureSetRequest.getSpec().getProject(),
+                                String.format("%s:%d", ref.getName(), ref.getVersion()))
+                            .inc());
+
+            featureRow.getFieldsList().stream()
+                .filter(field -> featureNames.containsKey(field.getName()))
+                .forEach(
+                    field -> {
+                      FeatureReference ref = featureNames.get(field.getName());
+                      String id = generateFeatureStringRef(ref);
+                      featureValuesMap.get(entityRow).put(id, field.getValue());
+                    });
+          }
+        }
+      }
+
+      List<FieldValues> fieldValues =
+          featureValuesMap.values().stream()
+              .map(valueMap -> FieldValues.newBuilder().putAllFields(valueMap).build())
+              .collect(Collectors.toList());
+      return getOnlineFeaturesResponseBuilder.addAllFieldValues(fieldValues).build();
+    }
+  }
+
+  @Override
+  public GetBatchFeaturesResponse getBatchFeatures(GetBatchFeaturesRequest getFeaturesRequest) {
+    throw Status.UNIMPLEMENTED.withDescription("Method not implemented").asRuntimeException();
```

Comment:
We could maybe use https://docs.oracle.com/javase/6/docs/api/java/lang/UnsupportedOperationException.html

Reply:
No, we should be using the GRPC statuses so that they are correctly thrown back to users.

Reply:
Ah I didn't realize there was only `asRuntimeException` and nothing more granular.